vendors/st/boards/stm32l475_discovery/ports/common_io/iot_spi.c

```diff
@@ -0,0 +1,679 @@
+                               uint8_t * const pvRxBuffer,
+                               size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: Add case for slave not selected? */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvTxBuffer == NULL ) || ( pvRxBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_TX_RX )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else if( HAL_SPI_TransmitReceive( pxSpi, pvTxBuffer, pvRxBuffer, ( uint16_t ) xBytes, IOT_SPI_BLOCKING_TIMEOUT ) != HAL_OK )
+        {
+            lError = IOT_SPI_TRANSFER_ERROR;
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_transfer_async( IotSPIHandle_t const pxSPIPeripheral,
+                                uint8_t * const pvTxBuffer,
+                                uint8_t * const pvRxBuffer,
+                                size_t xBytes )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    /* TODO: Add case for slave not selected? */
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else if( ( pvTxBuffer == NULL ) || ( pvRxBuffer == NULL ) || ( xBytes == 0 ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_GetState( pxSpi ) == HAL_SPI_STATE_BUSY_TX_RX )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else
+        {
+            HAL_NVIC_SetPriority( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum, 1, 0 );
+            HAL_NVIC_EnableIRQ( ( IRQn_Type ) ( ( STM32_SPI_HalContext_t * ) pxSPIPeripheral->pxSpiContext )->eIrqNum );
+
+            if( HAL_SPI_TransmitReceive_IT( pxSpi, pvTxBuffer, pvRxBuffer, ( uint16_t ) xBytes ) != HAL_OK )
+            {
+                lError = IOT_SPI_TRANSFER_ERROR;
+            }
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_spi_close( IotSPIHandle_t const pxSPIPeripheral )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    if( ( pxSPIPeripheral == NULL ) || ( pxSPIPeripheral->sOpened == IOT_SPI_CLOSED ) )
+    {
+        lError = IOT_SPI_INVALID_VALUE;
+    }
+    else
+    {
+        SPI_HandleTypeDef * pxSpi = pxSPIPeripheral->pxSpiContext->pxSpi;
+
+        if( HAL_SPI_Abort( pxSpi ) != HAL_OK )
+        {
+            lError = IOT_SPI_BUS_BUSY;
+        }
+        else if( HAL_SPI_DeInit( pxSpi ) != HAL_OK )
+        {
+            lError = IOT_SPI_BUS_BUSY;
```

Comment:
If this happens, is the app expected to retry this API call?

Reply:
After second check, HAL_SPI_DeInit doesn't fail. I removed this if condition.